elastic/elastic-agent
# Support flattened data_stream.* fields

## What does this PR do?

An input configuration supports flattened fields, however the 'data_stream' field was not being correctly decoded when flattened. This commit fixes this issue.

## Why is it important?

It increases consistency within the Elastic-Agent configuration and our other projects, like Beats, because they all support flattened configuration keys.

## Checklist

~~- [ ] I have made corresponding change to the default configuration files~~

~~## Author's Checklist~~

## How to test this PR locally
1. Deploy a standalone Elastic-Agent with a configuration like the following.
```yaml
outputs:
  default:
    type: elasticsearch
    hosts:
      - "${ELASTICSEARCH_HOST}"
    username: "${ELASTICSEARCH_USERNAME}"
    password: "${ELASTICSEARCH_PASSWORD}"

inputs:
  - type: filestream
    id: elastic-agent-input-id
    streams:
      - id: filestream-input-id-1
        data_stream:
          dataset: generic
        data_stream.namespace: "flattened-namespace"
        data_stream.type: "logs"
        paths:
          - /tmp/log.log
```
2. Make sure some fields from `data_stream.*` are flattened. It is even better if there is a mix
3. Look at the ingested data and make sure the fields correctly set.

## Related issues

- Closes https://github.com/elastic/elastic-agent/issues/3191
- Docs PR: https://github.com/elastic/ingest-docs/pull/489

~~## Use cases~~
~~## Screenshots~~
~~## Logs~~


## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
